# Microbiome data

```{r setup, echo=FALSE, results='asis'}
rebook::chapterPreamble()

suppressPackageStartupMessages({ 
    library(EnrichmentBrowser)
    library(curatedMetagenomicData)
    library(lefser)
    library(metagenomeSeq)
})
```

For abundance-based enrichment analysis on the **full abundance matrix**,
we recommend providing library-size normalized log counts for microbiome data.

When given raw read counts, we recommend to apply a variance-stabilizing 
transformation such as `voom` to arrive at library-size normalized
logCPMs.

Using the `EnrichmentBrowser::import` function it is also possible to import
fully processed metagenomic datasets and results of differential abundance (DA)
analysis from `r Biocpkg("limma")`, `r Biocpkg("edgeR")`, `r Biocpkg("DESeq2")`,
`r Biocpkg("lefser")`, and `r Biocpkg("metagenomeSeq")`.

As introduced in Section \@ref(sec:SE), the central data structure for working
with high-throughput experimental assay data in Bioconductor is the
`r Biocpkg("SummarizedExperiment")` container. 

In the following, we describe how to construct from scratch.
 
## Reading expression data from file 

```{r}
library(EnrichmentBrowser)
```

## `SummarizedExperiment` or `SingleCellExperiment` input

We also note that single-cell data stored in a `r Biocpkg("SingleCellExperiment")`

## Normalization

Normalization of high-throughput expression data is essential to make results
within and between samples comparable.
Microarray (intensity measurements) and RNA-seq (read counts) data typically
show distinct features that need to be normalized for.

The function `normalize` wraps commonly used functionality from
`r Biocpkg("limma")` for microarray normalization and from `r Biocpkg("EDASeq")`
for RNA-seq normalization.
For specific needs that deviate from these standard normalizations, the user
should always refer to more specific functions/packages.

Note that normalization for downstream DE analysis, e.g. with
`r Biocpkg("edgeR")` and `r Biocpkg("DESeq2")`, is not ultimately necessary
(and in some cases even discouraged) as many of these tools implement specific
normalization approaches themselves.
See the vignette of `r Biocpkg("EDASeq")`, `r Biocpkg("edgeR")`, and
`r Biocpkg("DESeq2")` for details.

Within-lane normalization to adjust for gene-specific effects such as gene
length and GC-content requires to retrieve this information first,
e.g. from BioMart or specific Bioconductor annotation packages.
Both modes are implemented in the `r Biocpkg("EDASeq")` function
`getGeneLengthAndGCContent`.

## Differential abundance analysis

The `r Biocpkg("EnrichmentBrowser")` incorporates established functionality from
the `r Biocpkg("limma")` package for differential expression analysis between sample
groups.
This involves the `voom`-transformation when applied to RNA-seq data.
Alternatively, differential expression analysis for RNA-seq data can also be
carried out based on the negative binomial distribution with `r Biocpkg("edgeR")`
and `r Biocpkg("DESeq2")`.

This can be performed using the function `deAna` and assumes some standardized
variable names:
 
- **GROUP** defines the sample groups being contrasted,
- **BLOCK** defines paired samples or sample blocks, as e.g. for batch effects.

For more information on experimental design, see the
[limma user's guide](https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf),
chapter 9.  

## Import from popular DA tools

### Import from `lefser`

```{r limma-emat}
library(limma)
library(edgeR)
group <- factor(se$GROUP)
design <- model.matrix(~ group)
keep <- filterByExpr(emat, design)
el <- voom(emat[keep,], design)
```

```{r limma-res}
fit <- lmFit(el, design)
fit <- eBayes(fit, robust = TRUE)
res <- topTable(fit, coef = 2, number = nrow(el), sort.by = "none")
```

```{r import-limma}
se <- EnrichmentBrowser::import(el, res)
```

### Import from `edgeR`

```{r edger-emat}
d <- DGEList(emat)
d <- calcNormFactors(d)
d <- estimateDisp(d, design)
```

```{r edger-res}
fit <- glmQLFit(d, design)
qlf <- glmQLFTest(fit)
res <- topTags(qlf, n = nrow(d), sort.by = "none")
```

```{r import-edger}
se <- EnrichmentBrowser::import(d, res)
se
```

### Import from `DESeq2`

```{r deseq-flow}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(emat,
                              colData = DataFrame(group = group),
                              design = ~ group)
dds <- DESeq(dds)
res <- results(dds)
```

```{r import-deseq}
se <- EnrichmentBrowser::import(dds, res)
se
```

## ID mapping

Using genomic information from different resources often requires mapping
between different types of gene identifiers.
Although primary analysis steps such as normalization and differential expression
analysis can be carried out independent of the gene ID type, downstream exploration
functionality of the `r Biocpkg("EnrichmentBrowser")` is consistently based on NCBI
Entrez Gene IDs.
It is thus, in this regard, beneficial to initially map gene IDs of a different
type to NCBI Entrez IDs.

The function `idTypes` lists the available ID types for the mapping
depending on the organism under investigation (here: _Homo sapiens_).

```{r idmap-idtypes}
idTypes("hsa")
```

## Case study

For microbiome data, we consider metagenomic sequencing data of fecal samples of 
colorectal carcinoma (CRC) patients

```{r zeller}
library(curatedMetagenomicData)
zeller.eset <- ZellerG_2014.metaphlan_bugs_list.stool()
exprs(eset)[1:5,1:5]
```

```{r}
counts <- sweep(exprs( zeller.eset ), 2, zeller.eset$number_reads / 100, "*")
counts <- round(counts)
exprs(zeller.eset) <- counts
```

```{r}        
zeller.eset <- zeller.eset[, zeller.eset$age < 60] 
zeller.eset <- zeller.eset[, zeller.eset$study_condition != "adenoma"]
table(zeller.eset$study_condition)
```

For RNA-seq data, we consider transcriptome profiles of four primary human
airway smooth muscle cell lines in two conditions: control and treatment with
dexamethasone
[Himes et al., 2014](https://doi.org/10.1371/journal.pone.0099625).

We load the
[airway](http://bioconductor.org/packages/airway)
dataset


```{R load-airway}
library(airway)
data(airway)
```

For further analysis, we only keep genes that are annotated to an ENSEMBL gene ID.

```{R preproc-airway}
airSE <- airway[grep("^ENSG", names(airway)), ]
dim(airSE)
assay(airSE)[1:4,1:4]
```

### Normalization

In case normalization is desired, between-lane normalization to adjust for
sequencing depth can be carried out as demonstrated for microarray data.

```{r norm-rseq}
airSE <- normalize(airSE, norm.method = "quantile")
```

TODO: GC + gene length normalization

### Differential expression analysis

For the airway dataset, it indicates whether the cell lines have been treated
with dexamethasone (1) or not (0).


```{R sample-groups-airway}
airSE$GROUP <- ifelse(colData(airway)$dex == "trt", 1, 0)
table(airSE$GROUP)
```

Paired samples, or in general sample batches/blocks, can be defined via a
`BLOCK` column in the `colData` slot. For the airway dataset, the sample blocks
correspond to the four different cell lines.


```{R sample-blocks}
airSE$BLOCK <- airway$cell
table(airSE$BLOCK)
```

For RNA-seq data, the `deAna` function can be used to carry out differential 
expression analysis between the two groups either based on functionality from
*limma* (that includes the `voom` transformation), or
alternatively, the frequently used *edgeR* or *DESeq2*
package. Here, we use the analysis based on *edgeR*.


```{R DE-ana-airway}
airSE <- deAna(airSE, de.method = "edgeR")
rowData(airSE)
```

### ID mapping

ID mapping for the airway dataset (from ENSEMBL to ENTREZ gene ids) can be
carried out using the function `idMap`.

```{r idmap-airway}
head(rownames(airSE))
airSE <- idMap(airSE, org = "hsa", from = "ENSEMBL", to = "ENTREZID")
head(rownames(airSE))
```

