# Microarray data

```{r setup, echo=FALSE, results='asis'}
rebook::chapterPreamble()

suppressPackageStartupMessages({ 
    library(EnrichmentBrowser)
    library(ALL)
    library(limma)
})
```

For expression-based enrichment analysis on the **full expression matrix**, 
we recommend providing normalized log2 intensities for microarray data.

Microarray data is expected to be single-channel. For two-color arrays, it is
expected that normalization within arrays has been already carried out,
e.g.~using `limma::normalizeWithinArrays`.

## Required inputs

1. A numeric *expression matrix*, in which the rows correspond to probes or 
   genes, and the columns to samples.
2. A numeric *group vector* containing a binary group assignment for the samples,
   using `0` and `1` for unaffected (controls) and affected (cases) samples,
   respectively.

For example, let's simulate an expression matrix for 100 probes and 6 microarrays
as described in the man pages of the `r Biocpkg("limma")` package.

In this simulated example, expression values are drawn from the normal distribution,

```{r}
sd <- 0.3 * sqrt(4 / rchisq(100, df = 4))
intens <- matrix(rnorm(100 * 6, sd = sd), 100, 6)
rownames(intens) <- paste("Gene", 1:100)
```

for two sample groups

```{r}
grp <- c(0, 0, 0, 1, 1, 1)
```

where the first two probes are differentially expressed in second group

```{r}
intens[1:2,4:6] <- intens[1:2,4:6] + 2
```

In this example, `intens` stores the expression matrix, and `grp` is the group vector.

## Import from `limma`

Create the expression data object
```{r}
library(limma)
```

1. 
```{r}
mdesign <- model.matrix(~factor(grp))
el <- new("EList", list(E = intens, design = mdesign))
```      
 
2. obtain differential expression results 
```{r}
fit <- lmFit(el, mdesign)
fit <- eBayes(fit, robust = TRUE) 
res <- topTable(fit, coef = 2, number = nrow(el), sort.by = "none")
```     

3. import
```{r}
se <- EnrichmentBrowser::import(el, res)
```

## Case study

For microarray data, we consider expression measurements of patients with acute
lymphoblastic leukemia
[Chiaretti et al., 2004](https://doi.org/10.1182/blood-2003-09-3243). A
frequent chromosomal defect found among these patients is a translocation, in
which parts of chromosome 9 and 22 swap places. This results in the oncogenic
fusion gene BCR/ABL created by positioning the ABL1 gene on chromosome 9 to a
part of the BCR gene on chromosome 22.

We load the
[ALL](http://bioconductor.org/packages/ALL)
dataset


```{R load-ALL}
library(ALL)
data(ALL)
```

and select B-cell ALL patients with and without the BCR/ABL fusion, as described previously
[Gentleman et al., 2005](https://www.bioconductor.org/help/publications/books/bioinformatics-and-computational-biology-solutions).


```{R subset-ALL}
ind.bs <- grep("^B", ALL$BT)
ind.mut <- which(ALL$mol.biol %in% c("BCR/ABL", "NEG"))
sset <- intersect(ind.bs, ind.mut)
all.eset <- ALL[, sset]
```

We can now access the expression values, which are intensity measurements
on a log-scale for 12,625 probes (rows) across 79 patients (columns).


```{R show-ALL}
dim(all.eset)
exprs(all.eset)[1:4,1:4]
```

### Preprocessing

As we often have more than one probe per gene, we compute gene expression values
as the average of the corresponding probe values.


```{R probe2gene, message = FALSE}
allSE <- probe2gene(all.eset)
head(names(allSE))
```

### Normalization

A default quantile normalization based on
`normalizeBetweenArrays` from `Biocpkg("limma")` can be carried out via

```{r norm-ma}
allSE <- normalize(allSE, norm.method = "quantile")
```

```{r plot-norm, fig.width=12, fig.height=6}
par(mfrow=c(1,2))
boxplot(assay(allSE, "raw"))
boxplot(assay(allSE, "norm"))
```

Note that this is only done for demonstration, as the ALL data has
been already RMA-normalized by the authors of the ALL dataset.

### Differential expression analysis

For the ALL dataset, the **GROUP** variable indicates whether the BCR-ABL gene
fusion is present (1) or not (0).

```{R sample-groups-ALL}
allSE$GROUP <- ifelse(allSE$mol.biol == "BCR/ABL", 1, 0)
table(allSE$GROUP)
```

For microarray expression data, the \Rfunction{deAna} function carries
out a differential expression analysis between the two groups based on
functionality from the \Biocpkg{limma} package.
Resulting fold changes and $t$-test derived $p$-values for each gene are
appended to the \Robject{rowData} slot.

```{r DE-ana-ALL}
allSE <- deAna(allSE, padj.method = "BH")
rowData(allSE)
```

Nominal _p_-values (`PVAL`) are corrected for multiple testing
(`ADJ.PVAL`) using the method from Benjamini and Hochberg implemented
in the function
[p.adjust](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/p.adjust.html) 
from the
[stats](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/00Index.html) package.

To get a first overview, we inspect the $p$-value distribution and the volcano
plot (fold change against $p$-value).

```{r plot-DE, fig.width=12, fig.height=6}
par(mfrow = c(1,2))
pdistr(rowData(allSE)$PVAL)
volcano(rowData(allSE)$FC, rowData(allSE)$ADJ.PVAL)
```

The expression change of highest statistical significance is observed for the
ENTREZ gene 7525.

```{r DE-exmpl}
ind.min <- which.min(rowData(allSE)$ADJ.PVAL)
rowData(allSE)[ind.min,]
```

This turns out to be the YES proto-oncogene 1
([hsa:7525@KEGG](http://www.genome.jp/dbget-bin/www_bget?hsa:7525)).

