# Microarray data

```{r setup, echo=FALSE}
suppressPackageStartupMessages({ 
    library(EnrichmentBrowser)
    library(ALL)
    library(limma)
})
```

For expression-based enrichment analysis on the _full expression matrix_, 
we recommend providing normalized log2 intensities for microarray data.

Microarray data is expected to be single-channel. For two-color arrays, it is
expected that normalization within arrays has been already carried out,
e.g.~using `limma::normalizeWithinArrays`.

## Required inputs

1. A numeric *expression matrix*, in which the rows correspond to probes or 
   genes, and the columns to samples.
2. A numeric *group vector* containing a binary group assignment for the samples,
   using `0` and `1` for unaffected (controls) and affected (cases) samples,
   respectively.

For example, let's simulate an expression matrix for 100 probes and 6 microarrays
as described in the man pages of the `r BiocPkg("limma")` package.

In this simulated example, expression values are drawn from the normal distribution,

```{r}
sd <- 0.3 * sqrt(4 / rchisq(100, df = 4))
intens <- matrix(rnorm(100 * 6, sd = sd), 100, 6)
rownames(intens) <- paste("Gene", 1:100)
```

for two sample groups

```{r}
grp <- c(0, 0, 0, 1, 1, 1)
```

where the first two probes are differentially expressed in second group

```{r}
y[1:2,4:6] <- y[1:2,4:6] + 2
```

In this example, `intens` stores the expression matrix, and `grp` is the group vector.

## Import from `limma`

Create the expression data object
```{r}
library(limma)
```

1. 
```{r}
el <- new("EList", list(E = intens, design = mdesign))
```      
 
2. obtain differential expression results 
```{r}
fit <- lmFit(el, mdesign)
fit <- eBayes(fit, robust = TRUE) 
res <- topTable(fit, coef = 2, number = nrow(el), sort.by = "none")
```     

3. import
```
se <- EnrichmentBrowser::import(el, res)
```

## Case study

For microarray data, we consider expression measurements of patients with acute
lymphoblastic leukemia
[Chiaretti et al., 2004](https://doi.org/10.1182/blood-2003-09-3243). A
frequent chromosomal defect found among these patients is a translocation, in
which parts of chromosome 9 and 22 swap places. This results in the oncogenic
fusion gene BCR/ABL created by positioning the ABL1 gene on chromosome 9 to a
part of the BCR gene on chromosome 22.

We load the
[ALL](http://bioconductor.org/packages/ALL)
dataset


```{R}
library(ALL)
data(ALL)
```

and select B-cell ALL patients with and without the BCR/ABL fusion, as described previously
[Gentleman et al., 2005](https://www.bioconductor.org/help/publications/books/bioinformatics-and-computational-biology-solutions).


```{R}
ind.bs <- grep("^B", ALL$BT)
ind.mut <- which(ALL$mol.biol %in% c("BCR/ABL", "NEG"))
sset <- intersect(ind.bs, ind.mut)
all.eset <- ALL[, sset]
```

```{R}
ind.bs <- grep("^B", ALL$BT)
ind.mut <- which(ALL$mol.biol %in% c("BCR/ABL", "NEG"))
sset <- intersect(ind.bs, ind.mut)
all.eset <- ALL[, sset]
```

We can now access the expression values, which are intensity measurements
on a log-scale for 12,625 probes (rows) across 79 patients (columns).


```{R}
dim(all.eset)
exprs(all.eset)[1:4,1:4]
```

As we often have more than one probe per gene, we compute gene expression values
as the average of the corresponding probe values.


```{R, message = FALSE}
allSE <- probe2gene(all.eset)
head(names(allSE))
```

